combine calls to run batch on unary requests and responses

apolcyn opened the pull request:
Combining run batch requests seems to have good improvement in unary call latency and qps. Testing on a single 8-core ubuntu host, single sequential unary latency seemed to improve by about 20%. Haven't seen benchmarks on different machines, but guessing it would be more noticeable in that case.

apolcyn commented:
test this please

murgatroid99 commented on src/ruby/lib/grpc/generic/active_call.rb:
```diff
@@ -226,6 +226,23 @@ def send_status(code = OK, details = '', assert_finished = false,
+      @send_initial_md_mutex.synchronize do
+        ops = {}
+        ops[SEND_INITIAL_METADATA] = @metadata_to_send unless @metadata_sent
+        @metadata_sent = true unless @metadata_sent
```
You don't need the "unless" here. `@metadata_sent` should be true after this line, one way or the other.

murgatroid99 commented on src/ruby/lib/grpc/generic/active_call.rb:
```diff
@@ -226,6 +226,23 @@ def send_status(code = OK, details = '', assert_finished = false,
     end
 
+    def server_unary_response(req, marshalled: false, trailing_metadata: {},
+                              code: Core::StatusCodes::OK, details: 'OK')
```
What is the point of the `marshalled` argument? I don't see any changes that use it, and I wouldn't expect users to pre-marshal data before passing it to us.

apolcyn commented:
latest commits change around streamer methods. work in progress to remove methods from active call that are now unused.

apolcyn commented:
the latest commits change around the streaming methods to also combine run batch calls where possible. it also removes so now-unused code of active call class.

this is still a work in progress, as there are still a couple of unaddressed comments.

apolcyn commented:
test this please

apolcyn commented:
Forgot to marshall calls on the modified streaming methods. The latest commits fix this, and also change around some error messages that were useful in debugging that.

I think this should be ready for another review.

apolcyn commented:
rebased and got rid of the submodule issues in https://github.com/grpc/grpc/pull/8511
